File: frontend/lib/schemas/staff-management.ts
import { z } from "zod";

// Mirrors backend/app/schemas/staff_management.py field-for-field.

export const EMPLOYMENT_STATUSES = ["active", "on_leave", "terminated"] as const;
export const ATTENDANCE_STATUSES = ["present", "absent", "leave", "half_day"] as const;

// Curated doc_type options for the upload UI (backend field is free text) —
// same allow-listed extensions as student documents (doc 13 feature 5:
// "same pattern as doc 07's student documents").
export const STAFF_DOCUMENT_TYPES = [
  { value: "contract", label: "Contract" },
  { value: "certification", label: "Certification" },
  { value: "id_document", label: "ID document" },
  { value: "other", label: "Other" },
] as const;
export const ALLOWED_DOCUMENT_EXTENSIONS = ["pdf", "jpg", "jpeg", "png"] as const;
export const ALLOWED_DOCUMENT_ACCEPT = ".pdf,.jpg,.jpeg,.png";

// ---------------------------------------------------------------- staff --

export const staffSchema = z.object({
  id: z.string(),
  user_id: z.string(),
  employee_no: z.string(),
  first_name: z.string(),
  last_name: z.string(),
  department: z.string(),
  designation: z.string(),
  qualification: z.string().nullable(),
  date_joined: z.string(),
  employment_status: z.string(),
  phone: z.string().nullable(),
  email: z.string().nullable(),
  is_active: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type Staff = z.infer<typeof staffSchema>;

export const staffCreateSchema = z.object({
  email: z.string().min(1, "Email is required").email("Enter a valid email address"),
  phone: z.string().optional().nullable(),
  first_name: z.string().min(1, "First name is required"),
  last_name: z.string().min(1, "Last name is required"),
  employee_no: z.string().min(1, "Employee number is required"),
  department: z.string().min(1, "Department is required"),
  designation: z.string().min(1, "Designation is required"),
  qualification: z.string().optional().nullable(),
  date_joined: z.string().min(1, "Date joined is required"),
  role_codes: z.array(z.string()).min(1, "Select at least one role"),
});
export type StaffCreate = z.infer<typeof staffCreateSchema>;

export const staffUpdateSchema = z.object({
  phone: z.string().optional().nullable(),
  email: z.string().email("Enter a valid email address").optional().nullable(),
  department: z.string().min(1).optional(),
  designation: z.string().min(1).optional(),
  qualification: z.string().optional().nullable(),
  employment_status: z.string().optional(),
});
export type StaffUpdate = z.infer<typeof staffUpdateSchema>;

// --------------------------------------------------------- assignments --

export const staffAssignmentSchema = z.object({
  id: z.string(),
  staff_id: z.string(),
  section_id: z.string(),
  academic_year_id: z.string(),
  term_id: z.string(),
  is_active: z.boolean(),
  created_at: z.string(),
});
export type StaffAssignment = z.infer<typeof staffAssignmentSchema>;

export const staffAssignmentCreateSchema = z.object({
  staff_id: z.string().min(1, "Select a teacher"),
  section_id: z.string().min(1, "Select a section"),
  academic_year_id: z.string().min(1, "Academic year is required"),
  term_id: z.string().min(1, "Term is required"),
});
export type StaffAssignmentCreate = z.infer<typeof staffAssignmentCreateSchema>;

// ----------------------------------------------------------- attendance --

export const staffAttendanceSchema = z.object({
  id: z.string(),
  staff_id: z.string(),
  date: z.string(),
  status: z.string(),
  check_in_time: z.string().nullable(),
  check_out_time: z.string().nullable(),
  marked_by: z.string().nullable(),
});
export type StaffAttendance = z.infer<typeof staffAttendanceSchema>;

// ------------------------------------------------------------ documents --

export const staffDocumentSchema = z.object({
  id: z.string(),
  staff_id: z.string(),
  doc_type: z.string(),
  file_url: z.string(),
  created_at: z.string(),
});
export type StaffDocument = z.infer<typeof staffDocumentSchema>;

// --------------------------------------------------------------- reports --

export const unassignedSectionRowSchema = z.object({
  section_id: z.string(),
  section_name: z.string(),
  class_name: z.string(),
});
export type UnassignedSectionRow = z.infer<typeof unassignedSectionRowSchema>;

export const unassignedTeacherRowSchema = z.object({
  staff_id: z.string(),
  first_name: z.string(),
  last_name: z.string(),
  employee_no: z.string(),
});
export type UnassignedTeacherRow = z.infer<typeof unassignedTeacherRowSchema>;

export const unassignedReportSchema = z.object({
  term_id: z.string().nullable(),
  unassigned_sections: z.array(unassignedSectionRowSchema),
  unassigned_teachers: z.array(unassignedTeacherRowSchema),
});
export type UnassignedReport = z.infer<typeof unassignedReportSchema>;

export const staffDirectoryRowSchema = z.object({
  id: z.string(),
  employee_no: z.string(),
  first_name: z.string(),
  last_name: z.string(),
  department: z.string(),
  designation: z.string(),
  employment_status: z.string(),
  phone: z.string().nullable(),
  email: z.string().nullable(),
  current_section_id: z.string().nullable(),
  current_section_name: z.string().nullable(),
  current_class_name: z.string().nullable(),
});
export type StaffDirectoryRow = z.infer<typeof staffDirectoryRowSchema>;

File: backend/app/schemas/staff_management.py
from datetime import date, datetime

from pydantic import BaseModel, EmailStr, Field

ATTENDANCE_STATUSES = {"present", "absent", "leave", "half_day"}
EMPLOYMENT_STATUSES = {"active", "on_leave", "terminated"}


# ---------------------------------------------------------------- staff --


class StaffCreate(BaseModel):
    """Staff onboarding (doc 13 feature 1): creates a `staff` row plus a
    linked `User` account via the invite-link flow (doc 04) — `role_codes`
    are the roles granted to the new account (defaults to `teacher`).
    """

    email: EmailStr
    phone: str | None = None
    first_name: str
    last_name: str
    employee_no: str
    department: str | None = None
    designation: str
    qualification: str | None = None
    date_joined: date
    role_codes: list[str] = Field(default_factory=lambda: ["teacher"])


class StaffUpdate(BaseModel):
    phone: str | None = None
    email: EmailStr | None = None
    department: str | None = None
    designation: str | None = None
    qualification: str | None = None
    employment_status: str | None = None


class StaffRead(BaseModel):
    id: str
    user_id: str
    employee_no: str
    first_name: str
    last_name: str
    department: str
    designation: str
    qualification: str | None
    date_joined: date
    employment_status: str
    phone: str | None
    email: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --------------------------------------------------------- assignments --


class StaffAssignmentCreate(BaseModel):
    staff_id: str
    section_id: str
    academic_year_id: str
    term_id: str


class StaffAssignmentRead(BaseModel):
    id: str
    staff_id: str
    section_id: str
    academic_year_id: str
    term_id: str
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ----------------------------------------------------------- attendance --


class StaffAttendanceEntry(BaseModel):
    staff_id: str
    date: date
    status: str
    check_in_time: datetime | None = None
    check_out_time: datetime | None = None


class StaffAttendanceBulkCreate(BaseModel):
    entries: list[StaffAttendanceEntry]


class StaffAttendanceRead(BaseModel):
    id: str
    staff_id: str
    date: date
    status: str
    check_in_time: datetime | None
    check_out_time: datetime | None
    marked_by: str | None

    model_config = {"from_attributes": True}


class StaffAttendanceRowResult(BaseModel):
    staff_id: str
    date: date
    success: bool
    error: str | None = None
    id: str | None = None


class StaffAttendanceBulkResult(BaseModel):
    results: list[StaffAttendanceRowResult]


# ------------------------------------------------------------ documents --


class StaffDocumentRead(BaseModel):
    id: str
    staff_id: str
    doc_type: str
    file_url: str
    created_at: datetime

    model_config = {"from_attributes": True}


# --------------------------------------------------------------- reports --


class UnassignedSectionRow(BaseModel):
    section_id: str
    section_name: str
    class_name: str


class UnassignedTeacherRow(BaseModel):
    staff_id: str
    first_name: str
    last_name: str
    employee_no: str


class UnassignedReport(BaseModel):
    """Operational checklist for the start of a term (doc 13) — not a
    workload-balancing report, since every teacher has exactly one class.
    """

    term_id: str | None
    unassigned_sections: list[UnassignedSectionRow]
    unassigned_teachers: list[UnassignedTeacherRow]


class StaffDirectoryRow(BaseModel):
    id: str
    employee_no: str
    first_name: str
    last_name: str
    department: str
    designation: str
    employment_status: str
    phone: str | None
    email: str | None
    current_section_id: str | None
    current_section_name: str | None
    current_class_name: str | None
